### ISSUE TITLE
Replace TEMPERDEV__* overridable functions with function pointers for easier customizability.

### ISSUE THREAD

**Issue by dangmoody** [dangmoody/Temper#14]
**TL;DR:** From a usability standpoint function pointers that are overridable are just better and come with significantly less friction for a performance cost that simply will not matter in the grand scheme of things.

-----

This system should be replaced with a struct full of function pointers.

Currently in Temper the way you override a function is done like so (example):

```c
#define TEMPER_IMPLEMENTATION
#define TEMPERDEV__ON_BEFORE_TEST OnBeforeTest
#define TEMPERDEV__ON_AFTER_TEST OnAfterTest
// and so on...
#include <temper/temper.h>
```

There are a few usability problems with this approach.

- In the above example, I would also need forward declarations of each function before the above code.  This is a somewhat arbitrary requirement vs. "Just set the pointer of those functions to the ones you have defined".
- Setting these overrides _must_ be done before including the main Temper header.  By using function pointers they would have to be overridden in the entry point where the rest of the Temper code _but_ it would ultimately provide greater flexibility for people who would be doing things beyond your most basic Temper entry point implementation.
- The defines on their own give absolutely no indication of what the function returns or what it's parameters are.  Every time you want to override one of these functions you end up having to look at the `#define` you want and then go find the corresponding default function, take the return type and parameters from that, and _then_ fill out your own function somewhere.  Function pointers would give you all of this information up front; removing friction.

When I initially implemented this system I wanted to mainly avoid having a struct full of function pointers and storing a pointer to that struct in the main temper test context.  I thought having a bunch of `#define`s would be just as simple to work with and I'd avoid having a bunch of function pointers taking 8 bytes each stored in a struct, ultimately leading to cache misses that I thought were otherwise avoidable.

However, the total cost of adding all of these function pointers will be far outweighed by the sheer usability improvements that come from it.  Temper will still run fast with this change in.
